### PR TITLE
don't accept inexact numbers while parsing date values, closes #180

### LINKED ISSUE #180: Denial of service parsing JsNumber as Instant/ZonedDateTime/LocalDateTime
### Play JSON Version (2.5.x / etc)

2.6.9

### API (Scala / Java / Neither / Both)

Scala 2.12.6

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

MacOS 10.13.6

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

Paste the output from `java -version` at the command line.

### Library Dependencies

N/A

### Expected Behavior

Run the following code:
```scala
  val number = BigDecimal(BigInt(Long.MaxValue), Int.MaxValue, MathContext.DECIMAL128)
  val jsNumber = JsNumber(number)
  val result = Reads.of[ZonedDateTime].reads(jsNumber)
```

It should return a result quickly.

### Actual Behavior

It is CPU bound for an unknown period of time (I have never waited long enough for it to complete).

### PR DESCRIPTION
**Pull request by OlegYch**
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Typesafe CLA](https://www.typesafe.com/contribute/cla)?
* [x] Have you [squashed your commits](https://www.playframework.com/documentation/latest/WorkingWithGit#Squashing-commits)?
* [x] Have you added copyright headers to new files?
* [x] Have you updated the documentation?
* [x] Have you added tests for any changed functionality?

## Fixes

Fixes #180 

## Purpose

Rejects inexact numbers as input for date parsers, they are limited to Long anyway

## Background Context

`BigDecimal.toLong` takes inappropriate amount of time for something like `1000000000e1000000000`

### CODE REVIEW

**Comment by OlegYch** [comment]
@dwijnand should i submit this against 2.6 branch too?

**Comment by dwijnand** [comment]
I'm not sure, but I would think we probably want this fixed in 2.6 too. Whether or not that means a second PR.. I also don't know. 🙂

**Comment by marcospereira** [comment]
Hi @OlegYch and @dwijnand,

Yes, we probably want to backport this to 2.6.x, but you won't need a new PR. We can cherry-pick on our side.

File: play-json/jvm/src/test/scala/play/api/libs/json/ReadsSpec.scala [hunk 1]
```diff
@@ -224,6 +224,12 @@ class ReadsSpec extends org.specs2.mutable.Specification {
       }
     }
 
+    "not be read from invalid number" in {
```
Comment (marcospereira) [inline play-json/jvm/src/test/scala/play/api/libs/json/ReadsSpec.scala]: Can we also have tests for `java.util.Date`, `java.sql.Date` and other affected?

File: play-json/shared/src/test/scala/play/api/libs/json/JsonSharedSpec.scala [hunk 2]
```diff
@@ -270,7 +270,7 @@ class JsonSharedSpec extends WordSpec
         "key3" -> js.arr(1, "tutu")
       )
 
-      js.prettyPrint(jo) mustEqual ("""{
+      js.prettyPrint(jo).replaceAllLiterally("\r\n", "\n") mustEqual ("""{
```
Comment (marcospereira) [inline play-json/shared/src/test/scala/play/api/libs/json/JsonSharedSpec.scala]: Why is this necessary now?

Comment (OlegYch) [reply play-json/shared/src/test/scala/play/api/libs/json/JsonSharedSpec.scala]: otherwise the test breaks on windows

File: play-json/jvm/src/test/scala/play/api/libs/json/ReadsSpec.scala (hunk 1 above)
Comment (OlegYch) [reply play-json/jvm/src/test/scala/play/api/libs/json/ReadsSpec.scala]: added more tests

File: play-json/shared/src/test/scala/play/api/libs/json/JsonSharedSpec.scala (hunk 2 above)
Comment (marcospereira) [reply play-json/shared/src/test/scala/play/api/libs/json/JsonSharedSpec.scala]: Oh, okay. Thanks for fixing that too. :-)
